File: src/electron/telemetry.ts


import { app } from "electron";
import { getConfig, setConfig } from "./config.js";
import { getSession } from "./auth.js";


import { API_URL } from "./lib/constants.js";


export const TELEMETRY_EVENTS = {
    APP_OPEN: "app_open",
    APP_CLOSE: "app_close",
    GAME_LAUNCH: "game_launch",
    GAME_CLOSE: "game_close",
    INSTANCE_CREATE: "instance_create",
    MOD_INSTALL: "mod_install",
    ERROR: "error",
} as const;

export type TelemetryEventType = typeof TELEMETRY_EVENTS[keyof typeof TELEMETRY_EVENTS];


let sessionStartTime: Date | null = null;
let gameStartTimes: Map<string, Date> = new Map();
let flushTimer: NodeJS.Timeout | null = null;
let isFlushing = false;

const telemetryQueue: Array<{
    eventType: TelemetryEventType;
    clientId: string;
    userId?: string;
    data?: Record<string, any>;
    launcherVersion: string;
    platform: string;
    locale: string;
}> = [];

const TELEMETRY_BATCH_SIZE = 20;
const TELEMETRY_FLUSH_INTERVAL_MS = 30 * 1000;
const TELEMETRY_MAX_QUEUE_SIZE = 500;
const TELEMETRY_USER_ID_CACHE_MS = 10 * 60 * 1000;
const TELEMETRY_USER_ID_RESOLVE_TIMEOUT_MS = 2000;

let cachedTokenUserId: string | null = null;
let cachedTokenValue: string | null = null;
let cachedTokenResolvedAt = 0;
let tokenUserIdResolvePromise: Promise<string | undefined> | null = null;

function extractCatIdUserId(uuid?: string | null): string | undefined {
    if (!uuid || !uuid.startsWith("catid-")) {
        return undefined;
    }

    const extracted = uuid.slice("catid-".length).trim();
    return extracted || undefined;
}


function getClientId(): string {
    const config = getConfig();

    
    if ((config as any).clientId) {
        return (config as any).clientId;
    }

    
    const clientId = crypto.randomUUID();

    
    setConfig({ clientId } as any);

    return clientId;
}


function isTelemetryEnabled(): boolean {
    const config = getConfig();
    return config.telemetryEnabled !== false; 
}


function getLauncherVersion(): string {
    return app.getVersion();
}


function getPlatform(): string {
    return process.platform; 
}


function getLocale(): string {
    const config = getConfig();
    return config.language || "th";
}


async function flushTelemetryQueue(force: boolean = false): Promise<void> {
    if (isFlushing) return;
    if (telemetryQueue.length === 0) return;

    if (flushTimer) {
        clearTimeout(flushTimer);
        flushTimer = null;
    }

    isFlushing = true;
    const batchSize = force ? telemetryQueue.length : Math.min(telemetryQueue.length, TELEMETRY_BATCH_SIZE);
    const events = telemetryQueue.splice(0, batchSize);

    try {
        const response = await fetch(`${API_URL}/telemetry/batch`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({ events }),
        });

        if (!response.ok) {
            throw new Error(`Batch send failed (${response.status})`);
        }
    } catch (error) {
        
        telemetryQueue.unshift(...events);
        if (telemetryQueue.length > TELEMETRY_MAX_QUEUE_SIZE) {
            telemetryQueue.splice(0, telemetryQueue.length - TELEMETRY_MAX_QUEUE_SIZE);
        }
        console.error("[Telemetry] Batch flush error:", error);
    } finally {
        isFlushing = false;
    }

    if (telemetryQueue.length > 0 && !force) {
        scheduleTelemetryFlush();
    }
}

function scheduleTelemetryFlush(): void {
    if (flushTimer) return;
    flushTimer = setTimeout(() => {
        flushTimer = null;
        void flushTelemetryQueue();
    }, TELEMETRY_FLUSH_INTERVAL_MS);
}

function getCachedTokenUserId(token?: string): string | undefined {
    if (!token) {
        return undefined;
    }

    if (cachedTokenValue !== token || !cachedTokenUserId) {
        return undefined;
    }

    if (Date.now() - cachedTokenResolvedAt > TELEMETRY_USER_ID_CACHE_MS) {
        return undefined;
    }

    return cachedTokenUserId;
}


export async function resolveTelemetryUserIdForSession(sessionOverride?: {
    uuid?: string;
    apiToken?: string;
} | null): Promise<string | undefined> {
    const session = sessionOverride || getSession();
    if (!session) {
        return undefined;
    }

    const fromUuid = extractCatIdUserId(session.uuid);
    if (fromUuid) {
        return fromUuid;
    }

    const token = session.apiToken?.trim();
    if (!token) {
        return undefined;
    }

    const cached = getCachedTokenUserId(token);
    if (cached) {
        return cached;
    }

    if (tokenUserIdResolvePromise && cachedTokenValue === token) {
        return tokenUserIdResolvePromise;
    }

    cachedTokenValue = token;
    tokenUserIdResolvePromise = (async () => {
        try {
            const controller = new AbortController();
            const timeout = setTimeout(() => controller.abort(), TELEMETRY_USER_ID_RESOLVE_TIMEOUT_MS);

            try {
                const response = await fetch(`${API_URL}/auth/session/me`, {
                    method: "GET",
                    headers: {
                        Authorization: `Bearer ${token}`,
                    },
                    signal: controller.signal,
                });

                if (!response.ok) {
                    return undefined;
                }

                const payload = await response.json() as { id?: string };
                const resolvedId = payload.id?.trim();
                if (!resolvedId) {
                    return undefined;
                }

                cachedTokenUserId = resolvedId;
                cachedTokenResolvedAt = Date.now();
                return resolvedId;
            } finally {
                clearTimeout(timeout);
            }
        } catch {
            return undefined;
        } finally {
            tokenUserIdResolvePromise = null;
        }
    })();

    return tokenUserIdResolvePromise;
}


function resolveTelemetryUserId(explicitUserId?: string): string | undefined {
    if (explicitUserId && explicitUserId.trim()) {
        return explicitUserId.trim();
    }

    const session = getSession();
    if (!session) {
        return undefined;
    }

    const fromUuid = extractCatIdUserId(session.uuid);
    if (fromUuid) {
        return fromUuid;
    }

    const token = session.apiToken?.trim();
    const cached = getCachedTokenUserId(token);
    if (cached) {
        return cached;
    }

    return undefined;
}


function queueEvent(
    eventType: TelemetryEventType,
    data?: Record<string, any>,
    userId?: string
): void {
    if (!isTelemetryEnabled()) {
        return;
    }

    const resolvedUserId = resolveTelemetryUserId(userId);
    if (!resolvedUserId) {
        
        void resolveTelemetryUserIdForSession();
    }

    telemetryQueue.push({
        eventType,
        clientId: getClientId(),
        userId: resolvedUserId,
        data: data || undefined,
        launcherVersion: getLauncherVersion(),
        platform: getPlatform(),
        locale: getLocale(),
    });

    if (telemetryQueue.length > TELEMETRY_MAX_QUEUE_SIZE) {
        telemetryQueue.splice(0, telemetryQueue.length - TELEMETRY_MAX_QUEUE_SIZE);
    }

    if (telemetryQueue.length >= TELEMETRY_BATCH_SIZE) {
        void flushTelemetryQueue();
    } else {
        scheduleTelemetryFlush();
    }
}






export function trackAppOpen(): void {
    sessionStartTime = new Date();

    const config = getConfig();
    const isFirstLaunch = !(config as any).hasLaunchedBefore;

    if (isFirstLaunch) {
        setConfig({ hasLaunchedBefore: true } as any);
    }

    queueEvent(TELEMETRY_EVENTS.APP_OPEN, {
        firstLaunch: isFirstLaunch,
    });
}


export function trackAppClose(): void {
    let sessionDuration = 0;

    if (sessionStartTime) {
        sessionDuration = Math.floor((Date.now() - sessionStartTime.getTime()) / 1000);
    }

    queueEvent(TELEMETRY_EVENTS.APP_CLOSE, {
        sessionDuration,
    });
}


export function trackGameLaunch(
    instanceId: string,
    version: string,
    loader?: string,
    userId?: string
): void {
    gameStartTimes.set(instanceId, new Date());

    queueEvent(TELEMETRY_EVENTS.GAME_LAUNCH, {
        instanceId,
        version,
        loader: loader || "vanilla",
    }, userId);
}


export function trackGameClose(instanceId: string, userId?: string): void {
    const startTime = gameStartTimes.get(instanceId);
    let playtime = 0;

    if (startTime) {
        playtime = Math.floor((Date.now() - startTime.getTime()) / 1000);
        gameStartTimes.delete(instanceId);
    }

    
    const now = Date.now();
    for (const [id, time] of gameStartTimes) {
        if (now - time.getTime() > 24 * 60 * 60 * 1000) {
            gameStartTimes.delete(id);
        }
    }

    queueEvent(TELEMETRY_EVENTS.GAME_CLOSE, {
        instanceId,
        playtime,
    }, userId);
}


export function trackInstanceCreate(
    version: string,
    loader: string,
    source?: string,
    userId?: string
): void {
    queueEvent(TELEMETRY_EVENTS.INSTANCE_CREATE, {
        version,
        loader,
        source: source || "manual",
    }, userId);
}


export function trackModInstall(
    modId: string,
    source: string,
    userId?: string
): void {
    queueEvent(TELEMETRY_EVENTS.MOD_INSTALL, {
        modId,
        source,
    }, userId);
}


export function trackError(
    message: string,
    context?: string,
    stack?: string
): void {
    queueEvent(TELEMETRY_EVENTS.ERROR, {
        message: message.substring(0, 500), 
        context,
        stack: stack ? stack.substring(0, 1000) : undefined, 
    });
}






export function initTelemetry(): void {
    console.log("[Telemetry] Initializing...");
    console.log("[Telemetry] Enabled:", isTelemetryEnabled());
    console.log("[Telemetry] Client ID:", getClientId());
    void resolveTelemetryUserIdForSession();

    
    trackAppOpen();
}


export async function cleanupTelemetry(): Promise<void> {
    console.log("[Telemetry] Cleaning up...");

    
    trackAppClose();
    await flushTelemetryQueue(true);
}

export default {
    initTelemetry,
    cleanupTelemetry,
    trackAppOpen,
    trackAppClose,
    trackGameLaunch,
    trackGameClose,
    trackInstanceCreate,
    trackModInstall,
    trackError,
    TELEMETRY_EVENTS,
};

File: src-tauri/src/telemetry.rs
use once_cell::sync::Lazy;
use serde::Serialize;
use std::collections::VecDeque;
use std::sync::Mutex;

const API_URL: &str = "https://api.reality.catlabdesign.space";
const TELEMETRY_BATCH_SIZE: usize = 20;
const TELEMETRY_FLUSH_INTERVAL_MS: u64 = 30_000;
const TELEMETRY_MAX_QUEUE_SIZE: usize = 500;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TelemetryEvent {
    pub event_type: String,
    pub client_id: String,
    pub user_id: Option<String>,
    pub data: Option<serde_json::Value>,
    pub launcher_version: String,
    pub platform: String,
    pub locale: String,
}

pub struct TelemetryQueue {
    events: VecDeque<TelemetryEvent>,
}

static QUEUE: Lazy<Mutex<TelemetryQueue>> = Lazy::new(|| {
    Mutex::new(TelemetryQueue {
        events: VecDeque::new(),
    })
});

static FLUSH_INITIALIZED: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));

// (api_token, users.id) — resolved from /auth/session/me for non-catid accounts.
// Keyed by token so an account switch invalidates it.
static TOKEN_USER_ID: Lazy<Mutex<Option<(String, String)>>> = Lazy::new(|| Mutex::new(None));

// When the app opened, used to compute the app_close session duration.
static SESSION_START: Lazy<Mutex<Option<std::time::Instant>>> = Lazy::new(|| Mutex::new(None));

fn get_client_id() -> String {
    let config = crate::config::config_get();
    let dir = config.minecraft_dir.unwrap_or_default();
    if dir.is_empty() {
        "unknown".to_string()
    } else {
        use std::hash::{Hash, Hasher};
        let mut hasher = std::collections::hash_map::DefaultHasher::new();
        dir.hash(&mut hasher);
        format!("{:x}", hasher.finish())
    }
}

fn get_platform() -> String {
    if cfg!(target_os = "windows") {
        "win32".to_string()
    } else if cfg!(target_os = "macos") {
        "darwin".to_string()
    } else {
        "linux".to_string()
    }
}

fn get_locale() -> String {
    crate::config::config_get()
        .extra
        .get("language")
        .and_then(|v| v.as_str())
        .unwrap_or("th")
        .to_string()
}

fn get_launcher_version() -> String {
    env!("CARGO_PKG_VERSION").to_string()
}

fn get_api_token() -> Option<String> {
    crate::auth::get_session_inner()?.api_token.clone()
}

fn resolve_user_id() -> Option<String> {
    let session = crate::auth::get_session_inner()?;
    if session.uuid.starts_with("catid-") {
        let id = session.uuid.strip_prefix("catid-")?.trim();
        if !id.is_empty() {
            return Some(id.to_string());
        }
    }
    // Microsoft/Minecraft accounts have no catid in the uuid — fall back to the
    // users.id resolved from the API token (see ensure_user_id_resolved). Without
    // this, telemetry.userId is null and presence/active-player counts never match.
    let token = session.api_token?;
    let cache = TOKEN_USER_ID.lock().ok()?;
    match cache.as_ref() {
        Some((cached_token, id)) if *cached_token == token => Some(id.clone()),
        _ => None,
    }
}

/// Resolve and cache users.id for non-catid accounts via /auth/session/me. Cheap
/// no-op once cached or for catid accounts. Call before presence-critical events.
pub async fn ensure_user_id_resolved() {
    let session = match crate::auth::get_session_inner() {
        Some(s) => s,
        None => return,
    };
    if session.uuid.starts_with("catid-") {
        return;
    }
    let token = match session.api_token {
        Some(t) => t,
        None => return,
    };

    if let Ok(cache) = TOKEN_USER_ID.lock() {
        if matches!(cache.as_ref(), Some((cached_token, _)) if *cached_token == token) {
            return;
        }
    }

    let resp = reqwest::Client::new()
        .get(format!("{API_URL}/auth/session/me"))
        .header("Authorization", format!("Bearer {token}"))
        .send()
        .await;

    if let Ok(resp) = resp {
        if resp.status().is_success() {
            if let Ok(json) = resp.json::<serde_json::Value>().await {
                if let Some(id) = json.get("id").and_then(|v| v.as_str()) {
                    let id = id.trim();
                    if !id.is_empty() {
                        if let Ok(mut cache) = TOKEN_USER_ID.lock() {
                            *cache = Some((token, id.to_string()));
                        }
                    }
                }
            }
        }
    }
}

async fn flush_telemetry_queue() {
    let events_to_send = {
        let mut queue = match QUEUE.lock() {
            Ok(q) => q,
            Err(_) => return,
        };
        if queue.events.is_empty() {
            return;
        }

        let batch_size = TELEMETRY_BATCH_SIZE.min(queue.events.len());
        let batch: Vec<TelemetryEvent> = queue.events.drain(..batch_size).collect();
        while queue.events.len() > TELEMETRY_MAX_QUEUE_SIZE {
            queue.events.pop_front();
        }
        batch
    };

    if events_to_send.is_empty() {
        return;
    }

    let body = serde_json::json!({ "events": events_to_send });
    let mut req = reqwest::Client::new()
        .post(format!("{API_URL}/telemetry/batch"))
        .header("Content-Type", "application/json");

    if let Some(token) = get_api_token() {
        req = req.header("Authorization", format!("Bearer {token}"));
    }

    match req.json(&body).send().await {
        Ok(resp) => {
            if !resp.status().is_success() {
                eprintln!("[Telemetry] Batch send failed: {}", resp.status());
                let mut queue = QUEUE.lock().unwrap();
                for event in events_to_send {
                    if queue.events.len() < TELEMETRY_MAX_QUEUE_SIZE {
                        queue.events.push_back(event);
                    }
                }
            }
        }
        Err(e) => {
            eprintln!("[Telemetry] Flush error: {e}");
            let mut queue = QUEUE.lock().unwrap();
            for event in events_to_send {
                if queue.events.len() < TELEMETRY_MAX_QUEUE_SIZE {
                    queue.events.push_back(event);
                }
            }
        }
    }
}

fn ensure_flush_timer() {
    let mut initialized = match FLUSH_INITIALIZED.lock() {
        Ok(i) => i,
        Err(_) => return,
    };
    if !*initialized {
        *initialized = true;
        drop(initialized);
        // tauri::async_runtime::spawn (not tokio::spawn) so this is safe to
        // trigger from the setup hook, which isn't inside a tokio context.
        tauri::async_runtime::spawn(async {
            loop {
                tokio::time::sleep(std::time::Duration::from_millis(
                    TELEMETRY_FLUSH_INTERVAL_MS,
                ))
                .await;
                flush_telemetry_queue().await;
            }
        });
    }
}

/// Force an out-of-band flush. Presence-critical events (game_launch/close) use
/// this so the active player count updates in seconds instead of waiting for the
/// 30s timer.
pub fn flush_now() {
    tauri::async_runtime::spawn(async { flush_telemetry_queue().await });
}

/// Emit `app_open` on every startup (parity with the Electron client, which the
/// Tauri port dropped). Without this, telemetry only records users who actually
/// launch a game — so version distribution / active-user stats were incomplete.
pub fn track_app_open() {
    *SESSION_START.lock().unwrap() = Some(std::time::Instant::now());

    // firstLaunch flag persists in config.extra so it's only true once ever.
    let mut config = crate::config::config_get();
    let first_launch = !config
        .extra
        .get("hasLaunchedBefore")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);
    if first_launch {
        config
            .extra
            .insert("hasLaunchedBefore".to_string(), serde_json::json!(true));
        let _ = crate::config::config_set(config);
    }

    queue_event("app_open", serde_json::json!({ "firstLaunch": first_launch }));
    flush_now();
}

/// Emit `app_close` with the session duration and drain the queue before exit.
/// Awaited from the ExitRequested handler so the final batch actually lands.
pub async fn track_app_close() {
    let session_duration = SESSION_START
        .lock()
        .unwrap()
        .map(|start| start.elapsed().as_secs())
        .unwrap_or(0);

    queue_event(
        "app_close",
        serde_json::json!({ "sessionDuration": session_duration }),
    );
    // Bound the final flush so a hung/offline network can't stall app exit.
    let _ = tokio::time::timeout(
        std::time::Duration::from_secs(3),
        flush_telemetry_queue(),
    )
    .await;
}

pub fn queue_event(event_type: &str, data: serde_json::Value) {
    ensure_flush_timer();

    let event = TelemetryEvent {
        event_type: event_type.to_string(),
        client_id: get_client_id(),
        user_id: resolve_user_id(),
        data: Some(data),
        launcher_version: get_launcher_version(),
        platform: get_platform(),
        locale: get_locale(),
    };

    let should_flush = {
        let mut queue = match QUEUE.lock() {
            Ok(q) => q,
            Err(_) => return,
        };
        queue.events.push_back(event);
        while queue.events.len() > TELEMETRY_MAX_QUEUE_SIZE {
            queue.events.pop_front();
        }
        queue.events.len() >= TELEMETRY_BATCH_SIZE
    };

    if should_flush {
        tauri::async_runtime::spawn(async { flush_telemetry_queue().await });
    }
}

#[tauri::command]
pub async fn telemetry_log_event(
    event_type: String,
    data: Option<serde_json::Value>,
) -> Result<(), String> {
    ensure_flush_timer();

    let event = TelemetryEvent {
        event_type,
        client_id: get_client_id(),
        user_id: resolve_user_id(),
        data,
        launcher_version: get_launcher_version(),
        platform: get_platform(),
        locale: get_locale(),
    };

    let should_flush = {
        let mut queue = QUEUE.lock().map_err(|e| e.to_string())?;
        queue.events.push_back(event);
        while queue.events.len() > TELEMETRY_MAX_QUEUE_SIZE {
            queue.events.pop_front();
        }
        queue.events.len() >= TELEMETRY_BATCH_SIZE
    };

    if should_flush {
        flush_telemetry_queue().await;
    }

    Ok(())
}
